File: src/main/java/galacticwars/clonewars/ability/AbilityDefinition.java
package galacticwars.clonewars.ability;

import java.util.Objects;

public record AbilityDefinition(
        AbilityId id,
        String displayName,
        AbilityKind kind,
        AbilityActivation activation,
        int cooldownTicks,
        int resourceCost,
        double range,
        int aiEvaluationIntervalTicks,
        boolean enabled
) {
    public AbilityDefinition {
        Objects.requireNonNull(id, "id");
        displayName = requireNonBlank(displayName, "displayName");
        Objects.requireNonNull(kind, "kind");
        Objects.requireNonNull(activation, "activation");
        requireNonNegative(cooldownTicks, "cooldownTicks");
        requireNonNegative(resourceCost, "resourceCost");
        if (!Double.isFinite(range) || range < 0.0D) {
            throw new IllegalArgumentException("range cannot be negative or non-finite");
        }
        if (aiEvaluationIntervalTicks < 1) {
            throw new IllegalArgumentException("aiEvaluationIntervalTicks must be positive");
        }
        if (kind == AbilityKind.FORCE && enabled) {
            throw new IllegalArgumentException("Force abilities are reserved but cannot be enabled yet");
        }
        if (activation == AbilityActivation.PASSIVE
                && (cooldownTicks != 0 || resourceCost != 0)) {
            throw new IllegalArgumentException("Passive abilities cannot consume resources or use cooldowns");
        }
    }

    public boolean active() {
        return activation != AbilityActivation.PASSIVE;
    }

    private static void requireNonNegative(int value, String label) {
        if (value < 0) {
            throw new IllegalArgumentException(label + " cannot be negative");
        }
    }

    private static String requireNonBlank(String value, String label) {
        Objects.requireNonNull(value, label);
        String normalized = value.trim();
        if (normalized.isEmpty()) {
            throw new IllegalArgumentException(label + " cannot be blank");
        }
        return normalized;
    }
}

File: src/main/kotlin/galacticwars/clonewars/ability/AbilityDefinition.kt
package galacticwars.clonewars.ability

class AbilityDefinition(
    id: AbilityId?,
    displayName: String?,
    kind: AbilityKind?,
    activation: AbilityActivation?,
    val cooldownTicks: Int,
    val resourceCost: Int,
    val range: Double,
    val aiEvaluationIntervalTicks: Int,
    val enabled: Boolean,
) {
    val id: AbilityId = requireNotNull(id) { "id" }
    val displayName: String = requireNonBlank(displayName, "displayName")
    val kind: AbilityKind = requireNotNull(kind) { "kind" }
    val activation: AbilityActivation = requireNotNull(activation) { "activation" }

    init {
        require(cooldownTicks >= 0) { "cooldownTicks cannot be negative" }
        require(resourceCost >= 0) { "resourceCost cannot be negative" }
        require(range.isFinite() && range >= 0.0) { "range cannot be negative or non-finite" }
        require(aiEvaluationIntervalTicks >= 1) { "aiEvaluationIntervalTicks must be positive" }
        require(activation != AbilityActivation.PASSIVE || cooldownTicks == 0 && resourceCost == 0) {
            "Passive abilities cannot consume resources or use cooldowns"
        }
    }

    fun id(): AbilityId = id

    fun displayName(): String = displayName

    fun kind(): AbilityKind = kind

    fun activation(): AbilityActivation = activation

    fun cooldownTicks(): Int = cooldownTicks

    fun resourceCost(): Int = resourceCost

    fun range(): Double = range

    fun aiEvaluationIntervalTicks(): Int = aiEvaluationIntervalTicks

    fun enabled(): Boolean = enabled

    fun active(): Boolean = activation != AbilityActivation.PASSIVE

    override fun equals(other: Any?): Boolean =
        this === other || other is AbilityDefinition &&
            id == other.id &&
            displayName == other.displayName &&
            kind == other.kind &&
            activation == other.activation &&
            cooldownTicks == other.cooldownTicks &&
            resourceCost == other.resourceCost &&
            range.compareTo(other.range) == 0 &&
            aiEvaluationIntervalTicks == other.aiEvaluationIntervalTicks &&
            enabled == other.enabled

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + displayName.hashCode()
        result = 31 * result + kind.hashCode()
        result = 31 * result + activation.hashCode()
        result = 31 * result + cooldownTicks
        result = 31 * result + resourceCost
        result = 31 * result + range.hashCode()
        result = 31 * result + aiEvaluationIntervalTicks
        result = 31 * result + enabled.hashCode()
        return result
    }

    override fun toString(): String =
        "AbilityDefinition[id=$id, displayName=$displayName, kind=$kind, activation=$activation, " +
            "cooldownTicks=$cooldownTicks, resourceCost=$resourceCost, range=$range, " +
            "aiEvaluationIntervalTicks=$aiEvaluationIntervalTicks, enabled=$enabled]"

    private companion object {
        fun requireNonBlank(value: String?, label: String): String {
            val normalized = requireNotNull(value) { label }.trim()
            require(normalized.isNotEmpty()) { "$label cannot be blank" }
            return normalized
        }
    }
}
